Multi-threading support: core, unix and cc3200

dpgeorge opened the pull request:
This PR adds multi-threading support to MicroPython in the form of the `_thread` module.  The work has been generously sponsored by Pycom (https://www.pycom.io).  The parts that make this up are:
- core: separating out the thread specific state from `mp_state_ctx` into an individual structure, accessed via `MP_STATE_THREAD`
- core: the NLR pointer is now per-thread (for x86, x86-64 and thumb archs only)
- core: thread-safe garbage collector (using a global mutex)
- core: thread-safe qstrs (using a global mutex)
- core: addition of py/mpthread.h: abstraction of threading primitive that need to be implemented by a given port
- core: addition of py/modthread.c to implement the `_thread` module in a port-agnostic way
- core: MICROPY_PY_THREAD option to enable the `_thread` module
- core: MICROPY_PY_THREAD_GIL option to enable the GIL (threading is still possible without the GIL but one must ensure not to access mutable structures (eg list) from different threads at the same time)
- tests: thread tests in tests/thread
- unix: implementation of threading using pthreads (see unix/mpthreadport.c)
- cc3200: implementation of threading using FreeRTOS threads (see cc3200/mpthreadport.c)

The GIL is not optimised (it releases/acquires on each pending exception check in the VM) but it works correctly.

To test on unix (requires pthread library):

```
cd unix
make
cd ../tests
./run-tests -d thread
```

dpgeorge committed:
tests: Add tests for _thread module.

Includes functionality and stress tests.

dpgeorge committed:
tests: Add 3 more tests for _thread module.

dpgeorge committed:
tests/thread: Remove need to sleep to wait for completion in some tests.

Use a lock and a counter instead, and busy wait for all threads to
complete.  This makes test run faster and they no longer rely on the time
module.

dpgeorge committed:
py: Add MP_STATE_THREAD to hold state specific to a given thread.

dpgeorge committed:
py: Add basic _thread module, with ability to start a new thread.

dpgeorge committed:
unix: Add basic thread support using pthreads.

Has the ability to create new threads.

dpgeorge committed:
py/modthread: Properly cast concrete exception pointer to an object.

dpgeorge committed:
py/modthread: Add stack_size() function.

dpgeorge committed:
py/modthread: Add exit() function.

Simply raises the SystemExit exception.

dpgeorge committed:
py/modthread: Implement lock object, for creating a mutex.

dpgeorge committed:
py/modthread: Add with-context capabilities to lock object.

dpgeorge committed:
py/gc: Make memory manager and garbage collector thread safe.

By using a single, global mutex, all memory-related functions (alloc,
free, realloc, collect, etc) are made thread safe.  This means that only
one thread can be in such a function at any one time.

dpgeorge committed:
tests/thread: Add tests for running GC within a thread, and heap stress.

dpgeorge committed:
py/modthread: Satisfy unused-args warning.

dpgeorge committed:
py/gc: Fix 2 cases of concurrent access to ATB and FTB.

dpgeorge committed:
unix/file: If write syscall returns because of EINTR then try again.

As per PEP-475.

dpgeorge committed:
py/modthread: Be more careful with root pointers when creating a thread.

dpgeorge committed:
py/modthread: Call mp_thread_start/mp_thread_finish around threads.

So the underlying thread implementation can do any necessary bookkeeping.

dpgeorge committed:
unix: Implement garbage collection with threading.

This patch allows any given thread to do a proper garbage collection and
scan all the pointers of all active threads.

dpgeorge committed:
py/gc: Fix GC+thread bug where ptr gets lost because it's not computed.

GC_EXIT() can cause a pending thread (waiting on the mutex) to be
scheduled right away.  This other thread may trigger a garbage
collection.  If the pointer to the newly-allocated block (allocated by
the original thread) is not computed before the switch (so it's just left
as a block number) then the block will be wrongly reclaimed.

This patch makes sure the pointer is computed before allowing any thread
switch to occur.

dpgeorge committed:
unix/mpthreadport: Use SA_SIGINFO for GC signal handler.

SA_SIGINFO allows the signal handler to access more information about
the signal, especially useful in a threaded environment.  The extra
information is not currently used but it may prove useful in the future.

dpgeorge committed:
unix/gccollect: Provide declaration of exported function.

dpgeorge committed:
unix/mpthreadport: Suppress compiler warning about unused arguments.

dpgeorge committed:
tests/thread: Rename thread_stress_XXX.py to stress_XXX.py.

dpgeorge committed:
tests/thread: Add tests that mutate shared objects.

Tests concurrent mutating access to: list, dict, set, bytearray.

dpgeorge committed:
tests/thread: Add test for concurrent interning of strings.

Qstr code accesses global state and needs to be made thread safe.

dpgeorge committed:
tests/thread: Add test for concurrent mutating of user instance.

dpgeorge committed:
py: Make interning of qstrs thread safe.

dpgeorge committed:
py: Implement a simple global interpreter lock.

This makes the VM/runtime thread safe, at the cost of not being able to
run code in parallel.

dpgeorge committed:
py: Don't use gc or qstr mutex when the GIL is enabled.

There is no need since the GIL already makes gc and qstr operations
atomic.

dpgeorge committed:
py/modthread: Make Lock objects work when GIL is enabled.

dpgeorge committed:
unix/modtime: Release the GIL when sleeping.

dpgeorge committed:
py/mpthread.h: Move GIL macros outside MICROPY_PY_THREAD block.

The GIL macros are needed even if threading is not enabled.

dpgeorge committed:
cc3200: Update FreeRTOS to v9.0.0.

This is a pristine copy (actually a subset of files) of upstream FreeRTOS
v9.0.0.

Modifications to the previous version of FreeRTOS (v8.1.2) included
addition of __attribute__ ((section (".boot"))) to the following
functions:

    pxPortInitialiseStack
    prvTaskExitError
    prvPortStartFirstTask
    xPortStartScheduler
    vPortSetupTimerInterrupt
    xTaskGenericCreate
    vTaskStartScheduler
    prvInitialiseTCBVariables
    prvInitialiseTaskLists
    prvAllocateTCBAndStack

This attribute will need to be reinstated on a case-by-case basis
because some of the above functions are now removed/changed.

dpgeorge committed:
cc3200: Define our own FreeRTOS heap so it can go in a special segment.

dpgeorge committed:
cc3200: Add basic threading capabilities.

Can create a new thread and run it.  Does not use the GIL at this point.

dpgeorge committed:
py/modthread: Allow to properly set the stack limit of a thread.

We rely on the port setting and adjusting the stack size so there is
enough room to recover from hitting the stack limit.

dpgeorge committed:
cc3200: Fix call to new exception to be _msg instead of _arg1.

dpgeorge committed:
py/nlrthumb: Convert NLR thumb funcs from asm to C with inline-asm.

Now only the bits that really need to be written in assembler are written
in it, otherwise C is used.  This means that the assembler code no longer
needs to know about the global state structure which makes it much easier
to maintain.

dpgeorge committed:
py/nlrsetjmp: Update to take into account new location of nlr_top.

It's now accessed via the MP_STATE_THREAD macro.

dpgeorge committed:
tests/thread: Make stack-size test run correctly and reliable on uPy.

dpgeorge committed:
tests/thread: Make sure that thread tests don't rely on floating point.

dpgeorge committed:
py/mpthread: Include mpstate.h when defining GIL macros.

dpgeorge committed:
cc3200: Use xTaskCreateStatic instead of osi_TaskCreate.

This allows to statically allocate the TCB (thread control block) and
thread stack in the BSS segment, reducing the need for dynamic memory
allocation.

dpgeorge committed:
cc3200/mpthreadport: Make mutex statically allocated.

Reduced the need for the FreeRTOS heap to allocate the mutex.

dpgeorge committed:
cc3200/mpthreadport: Properly initialise the main thread's data.

dpgeorge committed:
cc3200/gccollect: Use MP_STATE_THREAD(stack_top) to get top of stack.

dpgeorge committed:
cc3200/mpthreadport: Scan more root pointers from thread data.

dpgeorge committed:
cc3200/mpthreadport: Move mem alloc outside the thread_mutex lock.

Otherwise there could be a deadlock, with the GC's mutex and
thread_mutex.

dpgeorge committed:
cc3200: Enable the GIL.

dpgeorge committed:
tests/thread: Allow thread_sleep1 to run without floating point.

dpgeorge committed:
tests/thread: Allow some tests to run on ports with not much heap.

dpgeorge committed:
cc3200: In FreeRTOSConfig.h, comment on configSUPPORT_STATIC_ALLOCATION.

This config variable is now needed regardless of whether threading is
enabled or not.

dpgeorge committed:
cc3200: Remove .boot section attribute from osi_TaskCreate function.

This function is no longer used.  Having the .boot section attribute
meant that it was included in the firmware regargless of use.  Without
this attribute the linker can remove the function.

dpgeorge committed:
cc3200: Start the simplelink spawn task using the static task creator.

In VStartSimpleLinkSpawnTask we change xTaskCreate to xTaskCreateStatic
so that the task is created using statically allocated memory for the TCB
and stack.

This means that xTaskCreate function is no longer needed (the static
version is now used exclusively).

dpgeorge committed:
cc3200: Shrink the FreeRTOS heap and place TCB+stack in freed location.

The 16k FreeRTOS heap originally had all TCBs and stacks dynamically
allocated within it (plus semaphores and some other things).  Now that
xTaskCreateStatic is used instead of xTaskCreate, the TCBs and stacks
are allocated statically and no longer use any of the FreeRTOS heap.
Therefore, the FreeRTOS stack can be shrunk by the amount that has been
made static.  Furthermore, the TCBs and stack that are now static should
be placed in the .rtos_heaps section of RAM because this RAM is treated
specially by the bootloader (the bootloader executes from the first 16k
of RAM and loads the firmware into the section starting after the 16k).

After this patch the FreeRTOS heap (ucHeap) is 7200 bytes.  The memory
available for the MicroPython heap is 54936 bytes (including GC overhead).

stinos commented:
I was just looking into this a couple of days ago to eventually provide an msvc implementation. Seeing everything in the unix port is implemented on top of pthreads there are 2 ways: use an existing winapi-pthreads bridge or implement everything manually. The first seems the most interesting, except there are a couple of such implementations to choose from and not all of them are available as e.g. github repos so they would have to be copy-pasted. In any case: I don't have any use for threads myself so as long as there aren't any requests for this I'm probably not going to implement it.

pfalcon commented:
I kind of eyeballed this. Happy to confirm that desire to add VM* arg to every function doesn't come from this.

dpgeorge committed:
cc3200: Allow to compile bootloader with threading enabled.

Threading is not used in the bootloader but the config optios are still
enabled so we must exclude including FreeRTOS.h.

dpgeorge committed:
cc3200: Remove obsolete singleton heart-beat object.

dpgeorge committed:
qemu-arm: Disable gcc LTO option for nlrthumb.c.

If LTO is enabled for nlrthumb.c then gcc optimises away the nlr_push_tail
function when in fact it is needed.  So disable this option for this file.
